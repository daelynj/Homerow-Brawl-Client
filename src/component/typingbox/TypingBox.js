import React from "react";

function TypingBox(props) {
  function handleChange(event) {
    props.setStart(true);
    props.onChange(event.target.value);
  }

  function detectSpace(event) {
    var code = event.keyCode || event.which;
    if (code === 32 && checkWord()) {
      event.preventDefault();
    }
  }

  function checkWord() {
    if (props.value === props.word) {
      props.onChange("");
      props.setCurrentWordIndex(props.currentWordIndex + 1);
      return true;
    }
  }

  return (
    <div className="TypingBox">
      <input
        type="text"
        value={props.value}
        onChange={handleChange}
        onKeyDown={detectSpace}
      />
    </div>
  );
}

export default TypingBox;
