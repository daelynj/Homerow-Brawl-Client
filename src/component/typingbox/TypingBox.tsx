import * as React from "react";

interface Props {
  value: string;
  onChange: (newInput: string) => void;
  checkLetter: () => void;
  word: String;
  setCurrentWordIndex: (newIndex: number) => void;
  currentWordIndex: number;
  countUp: boolean;
}

export const TypingBox = (props: Props) => {
  const handleChange = (event: any) => {
    if (props.countUp === true) {
      props.onChange(event.target.value);
    }
  };

  const detectBackspace = (event: any) => {
    var code = event.keyCode || event.which;
    if (code !== 8) {
      props.checkLetter();
    }
  };

  const detectSpace = (event: any) => {
    var code = event.keyCode || event.which;
    if (code === 32 && checkWord()) {
      event.preventDefault();
    }
  };

  const checkWord = () => {
    if (props.value === props.word) {
      props.onChange("");
      props.setCurrentWordIndex(props.currentWordIndex + 1);
      return true;
    }
  };

  return (
    <div className="TypingBox">
      <input
        type="text"
        value={props.value}
        onKeyUp={detectBackspace}
        onKeyDown={detectSpace}
        onChange={handleChange}
      />
    </div>
  );
};
