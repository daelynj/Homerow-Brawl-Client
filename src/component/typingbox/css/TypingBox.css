.box {
  padding: 1%;
  margin: 1%;
  font-size: 20px;
  width: 90%;
  border: 5px solid goldenrod;
}
