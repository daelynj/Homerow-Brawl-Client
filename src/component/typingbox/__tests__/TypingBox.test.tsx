import * as React from "react";
import { TypingBox } from "../TypingBox";
import renderer from "react-test-renderer";
import { mount, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

configure({ adapter: new Adapter() });

describe("TypingBox", () => {
  const onChange = jest.fn();
  const setStart = jest.fn();
  const setCurrentWordIndex = jest.fn();
  const checkLetter = jest.fn();

  const buildProps = (newProps = {}) => ({
    value: "",
    onChange,
    setStart,
    word: "this",
    setCurrentWordIndex,
    currentWordIndex: 0,
    checkLetter,
    countUp: false,
    ...newProps
  });

  it("renders an input field", () => {
    const TypingBoxComponent = renderer
      .create(<TypingBox {...buildProps()} />)
      .toJSON();
    expect(TypingBoxComponent).toMatchSnapshot();
  });

  describe("when the countUp timer has not started", () => {
    it("renders the input field as empty when we pass a value", () => {
      const TypingBoxComponent = mount(<TypingBox {...buildProps()} />);

      TypingBoxComponent.find("input").simulate("change", {
        target: { value: "t" }
      });

      const inputValue = TypingBoxComponent.find("input").props().value;

      expect(inputValue).toEqual("");
    });

    describe("when we start typing", () => {
      const onChange = jest.fn();
      const TypingBoxComponent = mount(<TypingBox {...buildProps()} />);
      TypingBoxComponent.find("input").simulate("change", {
        target: { value: "t" }
      });

      it("does not call onChange", () => {
        expect(onChange).toHaveBeenCalledTimes(0);
      });
    });
  });

  describe("when the countUp timer is started", () => {
    it("renders the input field as empty when we pass an empty value", () => {
      const TypingBoxComponent = mount(
        <TypingBox {...buildProps({ countUp: true })} />
      );
      const inputValue = TypingBoxComponent.find("input").props().value;

      expect(inputValue).toEqual("");
    });

    describe("when we start typing", () => {
      const TypingBoxComponent = mount(
        <TypingBox {...buildProps({ countUp: true })} />
      );
      TypingBoxComponent.find("input").simulate("change", {
        target: { value: "t" }
      });

      it("calls onChange", () => {
        expect(onChange).toHaveBeenCalledWith("t");
      });
    });

    describe("when we press a space", () => {
      const TypingBoxComponent = mount(
        <TypingBox {...buildProps({ countUp: true })} />
      );
      TypingBoxComponent.find("input").simulate("change", {
        target: { value: " " }
      });

      it("calls onChange", () => {
        expect(onChange).toHaveBeenCalledWith(" ");
      });
    });

    describe("when we complete a word", () => {
      const TypingBoxComponent = mount(
        <TypingBox {...buildProps({ countUp: true })} />
      );

      TypingBoxComponent.find("input").simulate("change", {
        target: { value: "this" }
      });
      TypingBoxComponent.find("input").simulate("keyDown", {
        key: "Space",
        keyCode: 32,
        which: 32
      });

      it("renders the input field as empty", () => {
        expect(TypingBoxComponent.find("input").props().value).toEqual("");
      });
    });
  });
});
