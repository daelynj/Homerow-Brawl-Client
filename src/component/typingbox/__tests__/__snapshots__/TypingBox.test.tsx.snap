// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypingBox renders an input field 1`] = `
<input
  className="box"
  maxLength={11}
  onChange={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  type="text"
  value=""
/>
`;
